kernel/kernel-api/src/main/java/io/delta/kernel/commit/CommitMetadata.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (2025) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.kernel.commit;
+
+import static io.delta.kernel.internal.util.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import io.delta.kernel.annotation.Experimental;
+import io.delta.kernel.internal.actions.CommitInfo;
+import io.delta.kernel.internal.actions.Metadata;
+import io.delta.kernel.internal.actions.Protocol;
+import java.util.Optional;
+
+/**
+ * Contains all information (excluding the iterator of finalized actions) required to commit changes
+ * to a Delta table.
+ */
+@Experimental
+public class CommitMetadata {
+
+  private final long version;
+  private final String logPath;
+  private final CommitInfo commitInfo;
+  private final Optional<Protocol> readProtocolOpt;
+  private final Optional<Metadata> readMetadataOpt;
+  private final Optional<Protocol> newProtocolOpt;
+  private final Optional<Metadata> newMetadataOpt;
+
+  public CommitMetadata(
```

Comment:
aside: I wonder if we've discussed using something like [Immutables](https://immutables.github.io/) to reduce boiler plate